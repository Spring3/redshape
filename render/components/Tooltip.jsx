import React, { Component } from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';

const TooltipText = styled.p`
  position: absolute;
  display: none;
  border-radius: 3px;
  width: auto;
  padding: 5px;
  background: ${props => props.theme.bg};
  color: ${props => props.theme.mainDark};
  text-align: center;
  font-size: 12px;
  white-space: nowrap;
  box-shadow: 0px 2px 7px ${props => props.theme.minorText};
  z-index: 99;

  &::after {
    content: ' ';
    position: absolute;
    width: auto;    
    border: 2px solid black;
    border-width: 5px;
    border-color: ${props => props.theme.bg} transparent transparent transparent;
  }
  
  ${props => props.position === 'top' ? css`
    transform: translateX(-50%);
    // transform: translate(-50%, -50%);
    // top: calc(-250% + 5px);
    // top: -50px;
    bottom: 75%;
    left: 50%;
    
    &::after {
      top: 100%;
      left: 50%; 
      transform: translateX(-50%);
    }
  ` : css`
    left: 30px;
    transform: translateY(-50%);
    top: calc(-50% + 5px);
    
    &::after {
      top: calc(50% - 5px);
      left: -10px; 
      transform: rotate(90deg);
    }
  `}
`;

const TooltipContainer = styled.div`
  display: inline-block;
  position: relative;

  &:hover {
    ${TooltipText} {
      display: block;
    }
  }
`;

const TooltipMultiline = styled.div`
  display: flex;
  flex-direction: column;
`

class Tooltip extends Component {
  render() {
    const { className, children, text, position } = this.props;
    const lines = text.split('\\n');
    return (
      <TooltipContainer
        className={className}
      >
        {children}
        <TooltipText position={position || 'top'} className="tooltip">{ lines.length > 1 ? (<TooltipMultiline>{ lines.map(el => (<span>{el}</span>))}</TooltipMultiline>) : text }</TooltipText>
      </TooltipContainer>
    );
  }
}

Tooltip.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  text: PropTypes.string.isRequired,
  position: PropTypes.string,
};

Tooltip.defaultProps = {
  className: null
};

export default Tooltip;
