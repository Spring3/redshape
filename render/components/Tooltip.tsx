import React, { ReactNode } from 'react';
import { css } from '@emotion/react';
import { useTheme } from 'styled-components';

import { theme as Theme } from '../theme';

const styles = {
  container: css`
    display: inline-block;
    position: relative;

    &:hover {
      & > p {
        display: block;
      }
    }
  `,
  text: (theme: typeof Theme) => css`
    position: absolute;
    display: none;
    border-radius: 3px;
    top: -50px;
    left: 50%;
    width: auto;
    transform: translateX(-50%);
    padding: 5px;
    background: ${theme.bg};
    color: ${theme.mainDark};
    text-align: center;
    font-size: 12px;
    white-space: nowrap;

    &::after {
      content: ' ';
      position: absolute;
      top: 100%;
      left: 50%; 
      width: auto;    
      transform: translateX(-50%);
      border: 2px solid black;
      border-width: 5px;
      border-color: ${theme.bg} transparent transparent transparent;
    }
  `
};

type TooltipProps = {
  className?: string;
  children: ReactNode;
  text?: string;
};

const Tooltip = ({ className, children, text }: TooltipProps) => {
  const theme = useTheme() as typeof Theme;

  return (
    <div
      css={styles.container}
      className={className}
    >
      {children}
      <p css={styles.text(theme)} className="tooltip">{text}</p>
    </div>
  );
};

export {
  Tooltip
};
