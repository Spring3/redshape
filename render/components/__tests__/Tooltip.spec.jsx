import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Tooltip from '../Tooltip';

describe('Tooltip component', () => {
  afterEach(cleanup);

  it('should render the tooltip', () => {
    const { getByText } = render(
      <Tooltip className="container" text="Hello world">
        <i className="test" />
      </Tooltip>
    );
    expect(getByText('Hello world')).toBeDefined();
    expect(document.querySelector('i[class="test"]')).toBeDefined();
  });
});
