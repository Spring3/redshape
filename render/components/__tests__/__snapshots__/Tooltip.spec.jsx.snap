// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component should match the snapshot 1`] = `
.c2 {
  position: absolute;
  display: none;
  border-radius: 3px;
  width: auto;
  padding: 5px;
  text-align: center;
  font-size: 12px;
  white-space: nowrap;
  box-shadow: 0px 2px 7px;
  z-index: 99;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  bottom: 75%;
  left: 50%;
}

.c2::after {
  content: ' ';
  position: absolute;
  width: auto;
  border: 2px solid black;
  border-width: 5px;
  border-color: transparent transparent transparent;
}

.c2::after {
  top: 100%;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c0 {
  display: inline-block;
  position: relative;
}

.c0:hover .c1 {
  display: block;
}

<div
  className="container c0"
>
  <i
    className="test"
  />
  <p
    className="tooltip c1 c2"
  >
    Hello world
  </p>
</div>
`;
